src/mono/wasm/runtime/binding_support.js

```diff
@@ -1459,21 +1460,15 @@ var BindingSupportLib = {
 
 			var [delegateRoot, argsRoot] = MONO.mono_wasm_new_roots ([this.extract_mono_obj (delegate_obj), undefined]);
 			try {
-				if (!this.delegate_dynamic_invoke) {
-					if (!this.corlib)
-						this.corlib = this.assembly_load ("System.Private.CoreLib");
-					if (!this.delegate_class)
-						this.delegate_class = this.find_class (this.corlib, "System", "Delegate");
-					if (!this.delegate_class)
-					{
-						throw new Error("System.Delegate class can not be resolved.");
-					}
-					this.delegate_dynamic_invoke = this.find_method (this.delegate_class, "DynamicInvoke", -1);
-				}
-				argsRoot.value = this.js_array_to_mono_array (js_args);
-				if (!this.delegate_dynamic_invoke)
-					throw new Error("System.Delegate.DynamicInvoke method can not be resolved.");
-				return this.call_method (this.delegate_dynamic_invoke, delegateRoot.value, "m", [ argsRoot.value ]);
+				if (typeof delegate_obj.__mono_delegate_invoke__ === "undefined")
+					delegate_obj.__mono_delegate_invoke__ = this.mono_wasm_get_delegate_invoke(delegateRoot.value);
+				if (!delegate_obj.__mono_delegate_invoke__)
+					throw new Error("System.Delegate Invoke method can not be resolved.");
+
+				if (typeof delegate_obj.__mono_delegate_invoke_sig__ === "undefined")
+					delegate_obj.__mono_delegate_invoke_sig__ = Module.mono_method_get_call_signature (delegate_obj.__mono_delegate_invoke__, delegateRoot.value);
+
+				return this.call_method (delegate_obj.__mono_delegate_invoke__, delegateRoot.value, delegate_obj.__mono_delegate_invoke_sig__, js_args);
```

Comment:
Will `call_method` handle creating roots for the args? This removed `argsRoot.value = this.js_array_to_mono_array (js_args);`, so I'm wondering if that gets handled somewhere else now.

Also, if we are removing that, then `argsRoot` can be dropped completely.

Reply:
Addressed in recent commit.  Removed the argsRoot as it is no longer being used.

DynamicInvoke used an array of objects, thus the rooted object.  The Invoke method call is taken care of in the normal process of the invoke.

Reply:
call_method creates a result root but only for the duration of the call, you have to root it afterward.
args must be rooted explicitly because the actual arg list passed to the underlying method call impl is just some raw pointers and bits in the heap, but you don't need to necessarily root each arg if they're all in a managed array.